File: 8. atoi（自动机）/8. atoi.cpp
#include<cstdio>
#include<iostream>
#include<unordered_map>
#include<vector>
using namespace std;

class Automation{
	string state = "start";
	unordered_map<string, vector<string>> table = {
        {"start", {"start", "signed", "in_number", "end"}},
        {"signed", {"end", "end", "in_number", "end"}},
        {"in_number", {"end", "end", "in_number", "end"}},
        {"end", {"end", "end", "end", "end"}}
    };

public:
	int get_string(string str){
		// LeetCode don't allow overflow
		long long int ans = 0;
		int sign = 1;
		for (int i = 0; i < str.length(); i++){
			if(str[i] <= '9' && str[i] >= '0'){
				state = table[state][2];
			}
			else if(str[i] == '+'){
				state = table[state][1];
			}
			else if(str[i] == '-'){
				state = table[state][1];
			}
			else if(str[i] == ' '){
				state = table[state][0];
			}
			else{
				state = table[state][3];
			}
		
			if(state == "in_number"){
				if(sign*ans > 214748364 || sign*ans == 214748364 && str[i] - '0' > 7)
					return (1 << 31) - 1;
				else if(sign*ans < -214748364 || sign*ans == -214748364 && str[i] - '0' > 8)
					return -1 << 31;
				else
					ans = ans * 10 + str[i] - '0';
			}
			else if(state == "signed" && str[i] == '-'){
				sign = -1;
			}
		}
		return sign * ans;
	}
};

class Solution {
public:
    int myAtoi(string s) {
        Automation a;
	    return a.get_string(s);
    }
};

int main()
{
	Solution s;
	cout << s.myAtoi("21474836460");
	return 0;
}

File: 8. atoi（自动机）/8. atoi.py
class NFA:
        def __init__(self):
            self.ans = ""
            self.state = "start"
            self.table = {
                "start": ["start","signed","number","end"],
                "signed": ["end","end","number","end"],
                "number": ["end","end","number","end"],
                "end": ["end","end","end","end"]
            }

        def automation(self, string):
            for c in string:
                if c == ' ':
                    if self.state == "signed":
                        return 0
                    self.state = self.table[self.state][0]
                elif c == "+" or c == "-":
                    if self.state == "signed":
                        return 0
                    self.state = self.table[self.state][1]
                elif c.isdigit():
                    self.state = self.table[self.state][2]
                else:
                    if self.state == "signed":
                        return 0
                    self.state = self.table[self.state][3]
                if self.state == "signed" or self.state == "number":
                    self.ans += c
            if len(self.ans) == 0 or self.state == "signed":
                return 0
            else:
                return int(self.ans)

class Solution:   
    def myAtoi(self, s: str) -> int:
        nfa = NFA()
        ans = nfa.automation(s)
        if ans < -1 << 31:
            return -1 << 31
        elif ans > (1 << 31) - 1:
            return (1 << 31) - 1
        else:
            return ans
        
s = Solution()
print(s.myAtoi("-2345www"))